### ISSUE TITLE
Size limit on relay shares?

### ISSUE THREAD

**Issue by iankaufmann** [slskd/slskd#1228]
Could there possibly be a size limit on the amount of data that can be sent when adding shares from a relay agent node?

When I only share one small folder in my library (share size: 3.9 MB) with the controller, everything works fine:

```
[13:37:20 INF] Scanned 100% of shared directories. Found 60 files so far.
[13:37:20 INF] Scan found 60 files (and 0 were filtered) in 166ms
[13:37:20 INF] Removed or renamed 51861 files and 5323 directories in 36ms
[13:37:20 INF] Rebuilding filename index...
[13:37:20 INF] Version 0.21.4.0 is up to date.
[13:37:20 INF] Filename index rebuild complete
[13:37:20 INF] Shares scanned successfully. Found 7 directories and 60 files in 340.4526ms
[13:37:20 INF] Share cache loaded from disk successfully. Sharing 6 directories and 60 files
[13:37:20 INF] Warming browse response cache...
[13:37:20 INF] Browse response cached successfully in 14ms
[13:37:20 INF] Warming browse response cache...
[13:37:20 INF] Starting system clock...
[13:37:20 INF] Browse response cached successfully in 5ms
[13:37:20 INF] System clock started
[13:37:20 INF] Running in Agent relay mode; not connecting to the Soulseek server.
[13:37:20 INF] Attempting to connect to the relay controller https://x.x.x.x:5031/
[13:37:21 INF] Relay controller sent an authentication challenge
[13:37:21 INF] Relay controller connection established. Awaiting authentication...
[13:37:21 INF] Logging in...
[13:37:21 INF] Login succeeded.
[13:37:21 INF] Uploading shares...
[13:37:21 INF] Beginning upload of shares (3.9 MB)
[13:37:21 INF] Upload of shares succeeded (3.9 MB in 613ms)
[13:37:21 INF] Shares uploaded. Ready to relay files.
```

However, when I try to share my whole library (share size: 32.1 MB), I get this:

```
[13:41:20 INF] Scanned 100% of shared directories. Found 42122 files so far.
[13:41:20 INF] Scan found 42386 files (and 0 were filtered) in 15778ms
[13:41:20 INF] Removed or renamed 60 files and 6 directories in 3ms
[13:41:20 INF] Rebuilding filename index...
[13:41:20 INF] Filename index rebuild complete
[13:41:20 INF] Shares scanned successfully. Found 4252 directories and 42386 files in 16581.1753ms
[13:41:21 INF] Share cache loaded from disk successfully. Sharing 4251 directories and 42386 files
[13:41:22 INF] Warming browse response cache...
[13:41:22 INF] Browse response cached successfully in 1225ms
[13:41:23 INF] Warming browse response cache...
[13:41:23 INF] Starting system clock...
[13:41:23 INF] Browse response cached successfully in 1093ms
[13:41:23 INF] System clock started
[13:41:23 INF] Running in Agent relay mode; not connecting to the Soulseek server.
[13:41:23 INF] Attempting to connect to the relay controller https://x.x.x.x:5031/
[13:41:24 INF] Relay controller connection established. Awaiting authentication...
[13:41:24 INF] Relay controller sent an authentication challenge
[13:41:24 INF] Logging in...
[13:41:24 INF] Login succeeded.
[13:41:24 INF] Uploading shares...
[13:41:24 INF] Beginning upload of shares (32.1 MB)
[13:41:24 ERR] Error while copying content to a stream.
System.Net.Http.HttpRequestException: Error while copying content to a stream.
 ---> System.IO.IOException: Unable to write data to the transport connection: Broken pipe.
 ---> System.Net.Sockets.SocketException (32): Broken pipe
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Security.SslStream.<WriteSingleChunk>g__CompleteWriteAsync|157_1[TIOAdapter](ValueTask writeTask, Byte[] bufferToReturn)
   at System.Net.Security.SslStream.WriteAsyncChunked[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.IO.Stream.<CopyToAsync>g__Core|27_0(Stream source, Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
   at System.IO.Strategies.BufferedFileStreamStrategy.CopyToAsyncCore(Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.<CopyToAsync>g__WaitAsync|56_0(ValueTask copyTask)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpContent.<CopyToAsync>g__WaitAsync|56_0(ValueTask copyTask)
   at System.Net.Http.MultipartContent.SerializeToStreamAsyncCore(Stream stream, TransportContext context, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.<CopyToAsync>g__WaitAsync|56_0(ValueTask copyTask)
   at System.Net.Http.HttpConnection.SendRequestContentAsync(HttpRequestMessage request, HttpContentWriteStream stream, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at slskd.Relay.RelayClient.UploadSharesAsync(CancellationToken cancellationToken)
   at slskd.Relay.RelayClient.<StartAsync>b__57_0()
[13:41:24 ERR] Disconnecting from the relay controller
[13:41:24 INF] Relay controller connection disconnected
```

Curious if there is a parameter I can change that would allow this through?

The controller is on a Seedbox using the binary, and the agent is on UnRAID in a docker container.  I figured I was doing something wrong until I got it to work by simply sharing a smaller library.  That tells me that there isn't anything wrong with my .yml configurations or network.

**Comment by jpdillingham** [comment]
There's no size limit, but the upload does need to complete in one HTTP request in order to work properly.  If the network is unstable or slow this might be challenging.

I can look into adding retries if they don't exist, and will see if compression can help get the size down.

Unfortunately I think the relay feature might not be a great fit for low cost hosting solutions, and it's a shame because that's what I built it for.

**Comment by iankaufmann** [comment]
Yeah, it was just a guess... Seedbox is on a 40gbe connection and home is 1gbe, shouldn't be a speed issue.

The smaller share that I have been able to make is working great (1071 dirs / 9475 files).  I've served a ton of uploads in the last 24 hours off of that share.

I just can't figure out why when I share my bigger library I get the "Error while copying content to a stream" error almost immediately after it tries to start, and then it just keeps retrying over and over and failing each time until I stop it.

I thought maybe it could have been something else like special characters in folder names in the library (Björk vs Bjork) but fixing any of those hasn't helped either.

I had googled some of the errors I was seeing, like "System.IO.IOException: Unable to write data to the transport connection: Broken pipe", and saw this:
https://stackoverflow.com/questions/74005408/c-unable-to-write-data-to-the-transport-connection-broken-pipe-when-using-res

Now I know that thread probably has nothing to do with any of this but that is what made me ask if it was a size limit...

**Comment by jpdillingham** [comment]
Very interesting! I guess I shouldn't be surprised that there would be a built in request size limiter, as the ASP.NET framework is rather annoying with undocumented behavior.

Do you see anything in the logs on the controller side?

I can try and do some testing this weekend to see if I can reproduce the issue.

Also, the file that's being uploaded is an SQLite database, so special characters shouldn't be an issue.

Appreciate you digging!

**Comment by jpdillingham** [comment]
As it turns out, I _did_ know about this limitation because I had increased the limit to 10gb for file uploads.  #1229 adds the same increase for share uploads, and that should solve this issue.  Would really appreciate it if you could try again using the canary build (should be ready ~20 minutes from this post) and see if this resolves it.

In my defense I haven't looked at this code in 2+ years 😅

**Comment by jpdillingham** [comment]
Didn't mean to close this!

**Comment by iankaufmann** [comment]
No worries!  I can't believe that was it... I don't know anything about ASP.NET, but what I do know is how annoying it is to have people sending you support tickets all day long about things they think are broken and their wild theories about why.  Glad I wasn't way off!

Updated to canary on both ends and it's working perfectly.

This project is awesome and really solved my problem... which was being unwilling to open ports at home.  I'm not a leech anymore.  

Thanks for the quick fix!

**Comment by jpdillingham** [comment]
@iankaufmann thanks for the confirmation.

Closed by #1229
